### PR TITLE
IGNITE-18017 Add C++ tests for binary tuple

### PR DESCRIPTION
**Pull request by SammyVimes**
https://issues.apache.org/jira/browse/IGNITE-18017

### CODE REVIEW

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h [hunk 1]
```diff
@@ -431,7 +552,11 @@ class binary_tuple_builder {
      * @return Required size.
      */
     static SizeT gauge_number(const big_decimal &value) noexcept {
-        return SizeT(value.is_zero() ? 0 : value.get_unscaled_value().byte_size());
+        if (value.is_zero()) {
+            return 0;
+        }
+
+        return 4 /* scale */ + value.get_unscaled_value().byte_size();
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/binary_tuple_builder.h:559]: This violates Binary Tuple Format. We don't store decimal scale in binary tuple. It is stored in schema.

File: modules/platforms/cpp/ignite/schema/binary_tuple_parser.cpp [hunk 2]
```diff
@@ -251,6 +252,12 @@ big_integer binary_tuple_parser::get_number(bytes_view bytes) {
     return big_integer(bytes.data(), bytes.size());
 }
 
+big_decimal binary_tuple_parser::get_decimal(bytes_view bytes) {
+    auto scale = load_as<std::int32_t>(bytes);
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/binary_tuple_parser.cpp:256]: The scale should be provided as A method parameter. It is not stored in binary tuple. It is defined in schema.

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h (hunk 1 above)
Comment (SammyVimes) [reply modules/platforms/cpp/ignite/schema/binary_tuple_builder.h]: Right, fixed

File: modules/platforms/cpp/ignite/schema/binary_tuple_parser.cpp (hunk 2 above)
Comment (SammyVimes) [reply modules/platforms/cpp/ignite/schema/binary_tuple_parser.cpp]: Fixed

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h [hunk 3]
```diff
@@ -328,18 +446,21 @@ class binary_tuple_builder {
...
-        return (std::make_unsigned_t<T>(value) + std::numeric_limits<U>::max() + 1) <= std::numeric_limits<V>::max();
+    template <typename SRC, typename TGT>
+    static bool fits(SRC value) noexcept {
+        static_assert(std::is_signed_v<SRC>);
+        static_assert(std::is_signed_v<TGT>);
+        using U_TGT = std::make_unsigned_t<TGT>;
+
+        // Check if TGT::min <= value <= TGT::max.
+        return std::make_unsigned_t<SRC>(value + std::numeric_limits<TGT>::max() + 1)
+            <= std::numeric_limits<U_TGT>::max();
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/binary_tuple_builder.h:463]: Did you clang-format this? Also I think we can get rid of U_TGT.

File: modules/platforms/cpp/ignite/schema/binary_tuple_parser.cpp [hunk 4]
```diff
@@ -251,6 +252,11 @@ big_integer binary_tuple_parser::get_number(bytes_view bytes) {
     return big_integer(bytes.data(), bytes.size());
 }
 
+big_decimal binary_tuple_parser::get_decimal(bytes_view bytes, int32_t scale) {
+    big_integer mag(bytes.data(), bytes.size());
+    return {mag, scale};
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/binary_tuple_parser.cpp:257]: `std::move(mag)`

File: modules/platforms/cpp/ignite/schema/tuple_assembler.h [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+#include <type_traits>
+#include <vector>
+
+/**
+* This used to be the main tool to construct binary tuples (in IEP-54 format).
+* Now this is just a helper for unit tests.
+*/
+class tuple_assembler {
+private:
+    ignite::binary_tuple_schema schema;
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/tuple_assembler.h:35]: Please put the class itself into the `ignite` namespace and remove the `ignite::` prefix elsewhere.

File: modules/platforms/cpp/ignite/schema/tuple_assembler.h [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+    ignite::binary_tuple_builder builder;
+
+public:
+    /** Construct a new Tuple Assembler object. */
+    explicit tuple_assembler(ignite::binary_tuple_schema &sch)
+        : schema(std::move(sch))
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/tuple_assembler.h:42]: You have to define the parameter either as `binary_tuple_schema &&sch` or `binary_tuple_schema sch` for this to work properly.

File: modules/platforms/cpp/ignite/schema/tuple_assembler.h [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        , builder(schema.num_elements()) { }
+
+    /** Start a new tuple. */
+    void start() {
+        builder.start();
+    }
+
+    /** Append a null value. */
+    void appendNull() {
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/tuple_assembler.h:51]: 1. `append_null` 2. is it still used somewhere?

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.cpp [hunk 8]
```diff
@@ -269,9 +269,11 @@ void binary_tuple_builder::put_double(double value) {
 
 void binary_tuple_builder::put_number(bytes_view bytes) {
     big_integer value = binary_tuple_parser::get_number(bytes);
+    put_number(value);
+}
 
+void binary_tuple_builder::put_number(const ignite::big_integer &value) {
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/binary_tuple_builder.cpp:275]: Why use `ignite::` prefix here and below?

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h [hunk 9]
```diff
@@ -258,6 +314,68 @@ class binary_tuple_builder {
      */
     void append_uuid(uuid value) { put_uuid(value); }
 
+    /**
+     * @brief Appends a big integer for the next element.
+     *
+     * @param value Element value.
+     */
+    void append_number(const big_integer &value) { put_number(value); }
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/binary_tuple_builder.h:322]: Now I think there is no much point to have both `put_` and `append_` methods  essentially the same. We should lift the code from `put_` methods with specific type to corresponding `append_` methods.

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h [hunk 10]
```diff
@@ -431,7 +552,11 @@ class binary_tuple_builder {
      * @return Required size.
      */
     static SizeT gauge_number(const big_decimal &value) noexcept {
-        return SizeT(value.is_zero() ? 0 : value.get_unscaled_value().byte_size());
+        if (value.is_zero()) {
+            return 0;
+        }
+
+        return value.get_unscaled_value().byte_size();
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/binary_tuple_builder.h:559]: Hmm. What's the point of this change?

**Review by ademakov — CHANGES_REQUESTED** [review]

File: modules/platforms/cpp/ignite/schema/tuple_assembler.h (hunk 7 above)
Comment (SammyVimes) [reply modules/platforms/cpp/ignite/schema/tuple_assembler.h:51]: Only in a currently muted test

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h (hunk 10 above)
Comment (SammyVimes) [reply modules/platforms/cpp/ignite/schema/binary_tuple_builder.h:559]: Sorry, it's a leftover from the incorrect calculation

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h (hunk 3 above)
Comment (SammyVimes) [reply modules/platforms/cpp/ignite/schema/binary_tuple_builder.h:595]: Got it

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h (hunk 9 above)
Comment (SammyVimes) [reply modules/platforms/cpp/ignite/schema/binary_tuple_builder.h]: Done

File: modules/platforms/cpp/ignite/schema/tuple_assembler.h [hunk 11]
```diff
@@ -145,3 +142,4 @@ class tuple_assembler {
         return builder.build();
     }
 };
+}
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/tuple_assembler.h:145]: `} // ignite`

File: modules/platforms/cpp/ignite/schema/tuple_test.cpp [hunk 12]
```diff
@@ -0,0 +1,1268 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "tuple_assembler.h"
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <ignite/schema/binary_tuple_parser.h>
+#include <ignite/schema/column_info.h>
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/tuple_test.cpp:22]: The include order does not match our conventions.

Comment (SammyVimes) [reply modules/platforms/cpp/ignite/schema/tuple_test.cpp]: Fixed

File: modules/platforms/cpp/ignite/schema/binary_tuple_builder.h [hunk 13]
```diff
@@ -212,88 +268,282 @@ class binary_tuple_builder {
...
+     */
+    void append_timestamp(const ignite_timestamp &value);
+
+    /**
+     * @brief Appends a string for the next element.
+     *
+     * @param value Element value.
+     */
+    void append_string(const std::string &value) {
+        ignite::bytes_view bytes{reinterpret_cast<const std::byte *>(value.data()), value.size()};
```
Comment (ademakov) [inline modules/platforms/cpp/ignite/schema/binary_tuple_builder.h:544]: Again, why do you use the ignite prefix here?

**Review by ademakov — APPROVED** [review]

File: modules/platforms/cpp/ignite/schema/tuple_assembler.h (hunk 7 above)
Comment (ademakov) [reply modules/platforms/cpp/ignite/schema/tuple_assembler.h]: It's not clear how it is going to work now.

Comment (SammyVimes) [reply modules/platforms/cpp/ignite/schema/tuple_assembler.h]: Added an append

**Pull request merged**